src/backend/executor/nodeModifyTable.c

```diff
@@ -3186,6 +3186,11 @@ ExecInitModifyTable(ModifyTable *node, EState *estate, int eflags)
 	{
 		resultRelInfo = &mtstate->resultRelInfo[i];
 
+		if (resultRelInfo->ri_RelationDesc->rd_tableam &&
+			(table_scan_flags(resultRelInfo->ri_RelationDesc) & SCAN_FORCE_BIG_WRITE_LOCK))
+		{
+			LockRelation(resultRelInfo->ri_RelationDesc, ExclusiveLock);
```

Comment:
Hmm, when will unlock happen?

Reply:
The lock is transaction level, it will be unlocked at the end of transaction.